vassal-app/src/main/java/VASSAL/build/module/GameRefresher.java

```diff
@@ -689,25 +618,50 @@ private GamePiece processGamePiece(GamePiece piece, Command command) {
     return newPiece;
   }
 
-  private Point getDeckFreePosition(Deck deck) {
-    Point tempPosition = new Point(-1, -1);
-    Boolean correctTempPositionNotFound;
-    final GamePiece[] pieces;
-    pieces = deck.getMap().getAllPieces();
-    do {
-      correctTempPositionNotFound = false; //Assuming scan of pieces finds no match on tempPosition
-      for (final GamePiece piece : pieces) {
-        final Point piecePosition = piece.getPosition();
-        if (piecePosition.equals(tempPosition)) {
-          tempPosition.x -= 1;
-          correctTempPositionNotFound = true;
-          break;
-        }
-      }
-    } while (correctTempPositionNotFound);
-    return tempPosition;
-  }
+  /**
+   * Process the pieces in a Deck
+   * @param deck Deck to procss
+   * @param command Generated Command
+   */
+  private void processDeck(Deck deck, Command command) {
 
+    // Take a copy of the pieces in the Deck
+    final List<GamePiece> pieces = deck.asList();
+
+    // Remove all the pieces from the Deck by moving them to 0,0 then deleting them. This ensures the Undo
+    // ends with a valid MovePiece command to move the piece back into the Deck
+    for (final GamePiece piece : pieces) {
+      final Command move = deck.getMap().placeAt(piece, new Point(0, 0));
+      command.append(move);
+      final Command remove = new RemovePiece(piece);
+      remove.execute();
+      command.append(remove);
+    }
+
+    // Create a set of fresh pieces
+    final List<GamePiece> newPieces = new ArrayList<>();
+    for (final GamePiece piece : pieces) {
+      final GamePiece newPiece = gpIdChecker.createUpdatedPiece(piece);
```

Comment:
Deleting all the pieces and then adding them again makes me dizzy, like we might lose little chunks of state or something, but it _looks_ like it would work. 

Is the reason you don't just add e.g. "piece" back to the map (but instead create newPiece from it) something to do with the undo?

Reply:
This is the basic process of the Refresh as it has always worked, Remove the existing piece, Create a new piece, place the new piece back where the original piece was. This just pulls this all together for Decks rather than having Deck pieces refreshed in the standard processGamePiece loop.

Yes, this is possibly to ensure undo-ability as the new type and state of the piece has to be encoded into the Gamestate in  a Command that can be undone.